## Mais sofre funções e argumentos

### Revisão

In [183]:
#Função genérica
def nome(argumentos):
    #definição da função
    #return --> opcional
    pass
    

In [184]:
#Exemplo de função
def mostra_algo_na_tela():
    print('Isso é uma função')


mostra_algo_na_tela() #chamando a função


Isso é uma função


In [185]:
#Exemplo com arguemento e sem retorno
def mostra_algo_na_tela_arg(texto):
    print(texto)

mostra_algo_na_tela_arg('Texto necessário')

Texto necessário


In [186]:
#Exemplo com argumento e retorno
def perimetro_triangulo(a,b,c):
    perimetro = a + b + c
    
    return perimetro
    

perimetro_triangulo(3,4,5)
    
    

12

In [187]:
def perimetro_quadrilatero(a,b,c,d):
    perimetro = a + b + c + d
    
    return perimetro

perimetro_quadrilatero(3,4,5,6)

18

In [188]:
def perimetro_pentagono(a,b,c,d,e):
    perimetro = a + b + c + d + e
    
    return perimetro

perimetro_pentagono(3,4,5,6,7)

25

In [189]:
#Função que calcula o perímetro de qualquer polígono usando *args
def perimetro(*lados):
    perimetro = 0
    
    for lado in lados:
        perimetro = perimetro + lado
    
    return perimetro

perimetro(3,4,5,6,7,)

25

In [190]:
#Função que calcula o perímetro de qualquer polígono
#--> iterando pelo índice
def perimetro(*lados):
    perimetro = 0
    
    n = len(lados) #número de lados --> tamanho da tupla
    for i in range(n):
        perimetro += lados[i]
    
    return perimetro

print(perimetro(3,4,5,6,7))

25


In [191]:
#Exemplo do Lucas

lados = input('Digite os lados do poligono: ')
lados

Digite os lados do poligono: 1


'1'

In [192]:
#Função que calcula o perímetro de qualquer polígono
#--> iterando pelo índice
def perimetro(*lados):
    perimetro = 0
    
    n = len(lados) #número de lados --> tamanho da tupla
    for i in range(n):
        perimetro += lados[i]
    
    return perimetro


lista = (3, 4, 5, 6,7)

perimetro(*lista)

25

In [193]:
#Mostrando o uso do *args com entrada do usuário
def dados_usuarios(nome,DDD,*args):
    print(f'Notas de {nome} (ddd: {DDD}):')
    
    for nota in args:
        print(nota)

        
nome = input('Digite o nome:')
ddd = int(input('Digite o DDD:'))

n = int(input('Digite o numero de notas:'))

notas = []
for i in range(n):
    notas.append(int(input('Digite uma nota:')))

dados_usuarios(nome, ddd, *notas)

Digite o nome:1
Digite o DDD:1
Digite o numero de notas:1
Digite uma nota:1
Notas de 1 (ddd: 1):
1


In [194]:
#Outros aspectos de funções: argumentos posicionais (a ordem é importante)

def celular(numero, ddd = 11):
    print(f'({ddd}) {numero}')

celular('99876-5000',46)


(46) 99876-5000


In [205]:
# É possível mudar a ordem se o nome dos argumentos for utilizado explicitamente

celular(ddd=46,numero='99876-5000') #testando os argumentos invertidos

(46) 99876-5000


In [206]:
#Passando dicionários para funções: usando o **kwargs
def usuarios(**kwargs):
    for chave, valor in kwargs.items():
        print(chave, valor)


dados = {
    'nome':'Lucas',
    'ddd': 32,
}

usuarios(nome='Lucas',ddd=32, notaFinal = 9.2)

nome Lucas
ddd 32
notaFinal 9.2


In [207]:
#Passando dicionários para funções: construindo um dicionário
def usuarios(**kwargs):
    novo_dicionario = {}
    
    for chave, valor in kwargs.items():
        novo_dicionario[chave] = valor
    
    #print(novo_dicionario)
    
    return novo_dicionario
        



dados = usuarios(nome='Lucas',ddd=32, notaFinal = 9.2, peso = 80)
dados['nome']

'Lucas'

In [208]:
#Usando *args e **kwargs: para usarem ambos ao mesmo tempo, é necessário que o *args venha 
# antes do **kwargs

def  nome(a, b, c = 'f', *args, **kwargs):
    print(f'Tamanho do args: {len(args)}')


nome(1, 2,'f',5,4,True,'casa',sapato = 45)


Tamanho do args: 4


# Trabalhando com arquivos


In [209]:
#Abrindo um arquivo para leitura
# Modos de abertura: 
#--> r = read, w = write, a = append


try:
    arquivo = open('times.txt','r') # 'times.txt' é o arquivo que será aberto, 'r' = read
    
    conteudo = arquivo.read()
    
    print(conteudo)
    
    arquivo.close()

except FileNotFoundError:
    print('Arquivo nao encontrado!')
    


Time Pontos
Santos 11 
Gremio 02
Internacional 09   
Flamengo 09
Bahia 11


In [210]:
#Escrevendo em um arquivo novo

arq_escrita = open('alunos.txt','w') #o arquivo é criado ou sobrescrito (conteúdo anterior é perdido)

arq_escrita.write('Lucas ddd: 32\n')
arq_escrita.write('Amanda ddd: 11')

arq_escrita.close()

#Lendo o arquivo escrito
arq_leitura = open('alunos.txt','r')
conteudo = arq_leitura.read()
print(conteudo)
arq_leitura.close()
print("Fim da primeira leitura")



#Reescrevendo no arquivo 'alunos.txt'
arq_escrita = open('alunos.txt','w')

arq_escrita.write('Alexandre ddd: 11')

arq_escrita.close()



#Lendo o arquivo escrito
arq_leitura = open('alunos.txt','r')
conteudo = arq_leitura.read()
print(conteudo)
arq_leitura.close()



Lucas ddd: 32
Amanda ddd: 11
Fim da primeira leitura
Alexandre ddd: 11


In [211]:
#Escrevendo em um arquivo novo
arq_escrita = open('alunos.txt','w')

arq_escrita.write('Lucas ddd: 32\n')
arq_escrita.write('Amanda ddd: 11\n')

arq_escrita.close()

#Lendo o arquivo escrito
arq_leitura = open('alunos.txt','r')
conteudo = arq_leitura.read()
print(conteudo)
arq_leitura.close()
print("Fim da primeira leitura")



#Reescrevendo no arquivo 'alunos.txt' com o modo 'a' (append) -->cria o arquivo, caso não
# exista, ou acrescenta novas linhas no arquivo existente
arq_escrita = open('alunos.txt','a')

arq_escrita.write('Alexandre ddd: 11')

arq_escrita.close()



#Lendo o arquivo escrito
arq_leitura = open('alunos.txt','r')
conteudo = arq_leitura.read()
print(conteudo)
arq_leitura.close()



Lucas ddd: 32
Amanda ddd: 11

Fim da primeira leitura
Lucas ddd: 32
Amanda ddd: 11
Alexandre ddd: 11


In [212]:
#Abrindo um arquivo novo no modo 'a' (append) -->cria o arquivo, caso não
# exista, ou acrescenta novas linhas no arquivo existente
arq_escrita = open('atletas.txt','a')

arq_escrita.write('Alexandre carateca')

arq_escrita.close()


In [213]:
#Abrindo arquivos do tipo *.csv
import csv

arqcsv = open('dados_pessoais.csv','r')#abrindo para a leitura
planilha = csv.reader(arqcsv, delimiter = ';', lineterminator = '\n' )

for linha in planilha:
    print(linha)

arqcsv.close()



['Nome', 'Idade', 'Altura', 'Peso']
['Amanda', '20', '1.60', '50']
['Carlos', '21', '1.65', '72']
['Diego', '30', '1.82', '93']
['Fernanda', '16', '1.70', '67']
['Paulo', '28', '1.87', '70']
['Raquel', '42', '1.55', '80']


In [214]:
#Escrevendo em um arquivo CSV

import csv


arqcsv = open('minha_planilha.csv','w')

arq_escrita = csv.writer(arqcsv, delimiter = ' ', lineterminator = '\n')

arq_escrita.writerow(['Nome', 'Peso', 'Altura'])

arq_escrita.writerow(['Alexandre', '80', '1.85'])


arqcsv.close()




 - Exercício: fazer um programa que leia o arquivo 'dados.csv', cria e escreve outro arquivo 'dados_imc.csv' com as mesmas informações acrescentando uma coluna com o valor do IMC (índice de massa corpórea) de cada indivíduo.

- Tabela mostrando o arquivo 'dados.csv'

Nome | Idade | Altura |Peso |
---- | ------ |---- | ------
Amada | 20 | 1.60 | 50
Carlos|21|1.65|72
Diego|30|1.82|93
Fernanda|16|1.70|67
Paulo|28|1.87|70
Raquel|42|1.55|80


'dados.csv':
Nome;Idade;Altura;Peso
Amanda;20;1.60;50
Carlos;21;1.65;72
Diego;30;1.82;93
Fernanda;16;1.70;67
Paulo;28;1.87;70
Raquel;42;1.55;80

